gcloud-java-storage/src/main/java/com/google/gcloud/storage/testing/FakeStorageRpc.java

```diff
@@ -0,0 +1,264 @@
+package com.google.gcloud.storage.testing;
+
+import com.google.api.services.storage.model.Bucket;
+import com.google.api.services.storage.model.StorageObject;
+import com.google.gcloud.spi.ServiceRpcFactory;
+import com.google.gcloud.spi.StorageRpc;
+import com.google.gcloud.storage.StorageException;
+import com.google.gcloud.storage.StorageOptions;
+
+import sun.misc.IOUtils;
+import sun.reflect.generics.reflectiveObjects.NotImplementedException;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.math.BigInteger;
+import java.nio.file.FileAlreadyExistsException;
+import java.nio.file.NoSuchFileException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * A bare-bones in-memory implementation of Storage, meant for testing.
+ */
+public class FakeStorageRpc implements StorageRpc {
+
+  // fullname -> metadata
+  HashMap<String, StorageObject> stuff = new HashMap<>();
+  // fullname -> contents
+  HashMap<String, byte[]> contents = new HashMap<>();
+
+  private static final FakeStorageRpc instanceForTesting = new FakeStorageRpc();
+
+  // remove all files
+  public void reset() {
+    stuff = new HashMap<>();
+    contents = new HashMap<>();
+  }
+
+  // Returns StorageOptions that ask to use the static FakeStorageRpc instance,
+  // and resets it first so you start from a clean slate.
+  public static StorageOptions resetAndGetFakeStorageOptions() {
+    instanceForTesting.reset();
+    return StorageOptions.builder().serviceRpcFactory(new ServiceRpcFactory<StorageRpc,
+          StorageOptions>() {
+      @Override
+      public StorageRpc create(StorageOptions options) {
+        return instanceForTesting;
+      }
+    }).build();
+  }
+
+  @Override
+  public Bucket create(Bucket bucket, Map<Option, ?> options) throws StorageException {
+    return null;
```

Comment:
Without this (even in the most basic way) this utility will be not much help for most use-cases.

Can we implement it?  Simple implementation could just return bucket and bucket listing could iterate
over stuff and consolidate the one with the same bucket prefix (if you prefer not to create another datastructure
or further nest stuff)